### PR TITLE
[Merged by Bors] - chore: Delete `Algebra.Parity`

### PR DESCRIPTION
**Pull request by YaelDillies**
#12829 made `Algebra.Parity` almost empty. This PR deletes it.

---

- [x] depends on: #11986
- [x] depends on: #12829

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Group/Even.lean [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+* Try to generalize `IsSquare/Even` lemmas further. For example, there are still a few lemmas in
+  `Algebra.Ring.Parity` whose `Semiring` assumptions I (DT) am not convinced are necessary.
+* The "old" definition of `Even a` asked for the existence of an element `c` such that `a = 2 * c`.
+  For this reason, several fixes introduce an extra `two_mul` or `← two_mul`.
+  It might be the case that by making a careful choice of `simp` lemma, this can be avoided.
+
+## See also
+
+`Algebra.Ring.Parity` for the definition of odd elements.
```
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Algebra/Group/Even.lean]: I don't think this is useful but okay

**Comment by jcommelin** [comment]
|File|+-|Declaration|
|-|-|-|
| a/Mathlib/Algebra/Parity.lean | `-` | `def IsSquare (a : α) : Prop :=` |
| b/Mathlib/Algebra/Group/Even.lean | `+` | `def IsSquare (a : α) : Prop := ∃ r, a = r * r` |
| a/Mathlib/Algebra/Parity.lean | `-` | `def Odd (a : α) : Prop :=` |
| b/Mathlib/Algebra/Ring/Parity.lean | `+` | `def Odd (a : α) : Prop := ∃ k, a = 2 * k + 1` |
| b/Mathlib/Algebra/Group/Even.lean | `+` | `instance Additive.instDecidablePredEven [DecidablePred (IsSquare : α → Prop)] :` |
| a/Mathlib/Algebra/Parity.lean | `-` | `instance [DecidablePred (Even : α → Prop)] : DecidablePred (IsSquare : Multiplicative α → Prop) :=` |
| a/Mathlib/Algebra/Parity.lean | `-` | `instance [DecidablePred (IsSquare : α → Prop)] : DecidablePred (Even : Additive α → Prop) :=` |
| b/Mathlib/Algebra/Group/Even.lean | `+` | `instance Multiplicative.instDecidablePredIsSquare [DecidablePred (Even : α → Prop)] :` |
| b/Mathlib/Algebra/Ring/Parity.lean | `+` | `lemma Dvd.dvd.even (hab : a ∣ b) (ha : Even a) : Even b := ha.trans_dvd hab` |
| b/Mathlib/Algebra/Ring/Parity.lean | `+` | `lemma Even.add_odd : Even a → Odd b → Odd (a + b) := by` |
| b/Mathlib/Algebra/Ring/Parity.lean | `+` | `lemma even_iff_exists_two_mul : Even a ↔ ∃ b, a = 2 * b := by simp [even_iff_exists_two_nsmul]` |
| b/Mathlib/Algebra/Ring/Parity.lean | `+` | `lemma even_iff_two_dvd : Even a ↔ 2 ∣ a := by simp [Even, Dvd.dvd, two_mul]` |
| b/Mathlib/Algebra/Group/Even.lean | `+` | `lemma Even.isSquare_zpow [Group α] {n : ℤ} : Even n → ∀ a : α, IsSquare (a ^ n) := by` |
| b/Mathlib/Algebra/Ring/Parity.lean | `+` | `lemma Even.neg_one_pow (h : Even n) : (-1 : α) ^ n = 1 := by rw [h.neg_pow, one_pow]` |
| b/Mathlib/Algebra/Ring/Parity.lean | `+` | `lemma Even.neg_one_zpow (h : Even n) : (-1 : α) ^ n = 1 := by rw [h.neg_zpow, one_zpow]` |
| b/Mathlib/Algebra/Ring/Parity.lean | `+` | `lemma even_neg_two : Even (-2 : α) := by simp only [even_neg, even_two]` |
| b/Mathlib/Algebra/Ring/Parity.lean | `+` | `lemma Even.neg_zpow : Even n → ∀ a : α, (-a) ^ n = a ^ n := by` |
| b/Mathlib/Algebra/Ring/Parity.lean | `+` | `lemma Even.odd_add (ha : Even a) (hb : Odd b) : Odd (b + a) := add_comm a b ▸ ha.add_odd hb` |
| b/Mathlib/Algebra/Group/Even.lean | `+` | `lemma even_ofMul_iff {a : α} : Even (Additive.ofMul a) ↔ IsSquare a := Iff.rfl` |
| b/Mathlib/Algebra/Ring/Parity.lean | `+` | `lemma Even.pow_of_ne_zero (ha : Even a) : ∀ {n : ℕ}, n ≠ 0 → Even (a ^ n)` |
| b/Mathlib/Algebra/Ring/Parity.lean | `+` | `lemma Even.sub_odd (ha : Even a) (hb : Odd b) : Odd (a - b) := by` |
| b/Mathlib/Algebra/Group/Even.lean | `+` | `lemma even_toAdd_iff {a : Multiplicative α} : Even (Multiplicative.toAdd a) ↔ IsSquare a := Iff.rfl` |
| b/Mathlib/Algebra/Ring/Parity.lean | `+` | `lemma Even.trans_dvd (ha : Even a) (hab : a ∣ b) : Even b :=` |
| b/Mathlib/Algebra/Order/Sub/Canonical.lean | `+` | `lemma Even.tsub [CanonicallyLinearOrderedAddCommMonoid α] [Sub α] [OrderedSub α]` |
| b/Mathlib/Algebra/Ring/Parity.lean | `+` | `lemma even_two_mul (a : α) : Even (2 * a) := ⟨a, two_mul _⟩` |
| b/Mathlib/Algebra/Group/Even.lean | `+` | `lemma IsSquare.div [DivisionCommMonoid α] {a b : α} (ha : IsSquare a) (hb : IsSquare b) :` |
| b/Mathlib/Algebra/Group/Even.lean | `+` | `lemma isSquare_iff_exists_sq (m : α) : IsSquare m ↔ ∃ c, m = c ^ 2 := by simp [IsSquare, pow_two]` |
| b/Mathlib/Algebra/Group/Even.lean | `+` | `lemma IsSquare.map [MulOneClass α] [MulOneClass β] [FunLike F α β] [MonoidHomClass F α β]` |
| b/Mathlib/Algebra/Group/Even.lean | `+` | `lemma IsSquare.mul [CommSemigroup α] {a b : α} : IsSquare a → IsSquare b → IsSquare (a * b) := by` |
| b/Mathlib/Algebra/Group/Even.lean | `+` | `lemma isSquare_one [MulOneClass α] : IsSquare (1 : α) := ⟨1, (mul_one _).symm⟩` |
| b/Mathlib/Algebra/Group/Even.lean | `+` | `lemma isSquare_op_iff {a : α} : IsSquare (op a) ↔ IsSquare a :=` |
| b/Mathlib/Algebra/Group/Even.lean | `+` | `lemma isSquare_toMul_iff {a : Additive α} : IsSquare (Additive.toMul a) ↔ Even a := Iff.rfl` |
| b/Mathlib/Algebra/Group/Even.lean | `+` | `lemma isSquare_unop_iff {a : αᵐᵒᵖ} : IsSquare (unop a) ↔ IsSquare a := isSquare_op_iff.symm` |
| b/Mathlib/Algebra/Ring/Parity.lean | `+` | `lemma Odd.add_even (ha : Odd a) (hb : Even b) : Odd (a + b) := add_comm a b ▸ hb.add_odd ha` |
| b/Mathlib/Algebra/Ring/Parity.lean | `+` | `lemma Odd.add_odd : Odd a → Odd b → Even (a + b) := by` |
| b/Mathlib/Algebra/Ring/Parity.lean | `+` | `lemma odd_iff_exists_bit1 : Odd a ↔ ∃ b, a = bit1 b := exists_congr fun b ↦ by rw [two_mul]; rfl` |
| b/Mathlib/Algebra/Ring/Parity.lean | `+` | `lemma Odd.map [FunLike F α β] [RingHomClass F α β] (f : F) : Odd a → Odd (f a) := by` |
| b/Mathlib/Algebra/Ring/Parity.lean | `+` | `lemma Odd.neg (hp : Odd a) : Odd (-a) := by` |
| b/Mathlib/Algebra/Ring/Parity.lean | `+` | `lemma odd_neg_one : Odd (-1 : α) := by simp` |
| b/Mathlib/Algebra/Ring/Parity.lean | `+` | `lemma Odd.neg_pow : Odd n → ∀ a : α, (-a) ^ n = -a ^ n := by` |
| b/Mathlib/Algebra/Order/Ring/Canonical.lean | `+` | `lemma Odd.pos [Nontrivial α] : Odd a → 0 < a := by rintro ⟨k, rfl⟩; simp [pos_iff_ne_zero]` |
| b/Mathlib/Algebra/Ring/Parity.lean | `+` | `lemma Odd.pow (ha : Odd a) : ∀ {n : ℕ}, Odd (a ^ n)` |
| b/Mathlib/Algebra/Ring/Parity.lean | `+` | `lemma Odd.sub_even (ha : Odd a) (hb : Even b) : Odd (a - b) := by` |
| b/Mathlib/Algebra/Ring/Parity.lean | `+` | `lemma Odd.sub_odd (ha : Odd a) (hb : Odd b) : Even (a - b) := by` |
| b/Mathlib/Algebra/Ring/Parity.lean | `+` | `lemma odd_two_mul_add_one (a : α) : Odd (2 * a + 1) := ⟨_, rfl⟩` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem Dvd.dvd.even (hab : a ∣ b) (ha : Even a) : Even b :=` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem even_abs [AddGroup α] [LinearOrder α] {a : α} : Even |a| ↔ Even a := by` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem Even.add_odd : Even a → Odd b → Odd (a + b) := by` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem even_iff_exists_bit0 [Add α] {a : α} : Even a ↔ ∃ b, a = bit0 b :=` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem even_iff_exists_two_mul (a : α) : Even a ↔ ∃ b, a = 2 * b := by` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem even_iff_two_dvd : Even a ↔ 2 ∣ a := by simp [Even, Dvd.dvd, two_mul]` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem Even.isSquare_pow : Even n → ∀ a : α, IsSquare (a ^ n) := by` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem Even.isSquare_zpow [Group α] {n : ℤ} : Even n → ∀ a : α, IsSquare (a ^ n) := by` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem Even.mul_left (ha : Even a) (b) : Even (b * a) :=` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem Even.mul_right (ha : Even a) (b) : Even (a * b) :=` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem Even.neg_one_pow (h : Even n) : (-1 : α) ^ n = 1 := by rw [h.neg_pow, one_pow]` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem Even.neg_one_zpow (h : Even n) : (-1 : α) ^ n = 1 := by rw [h.neg_zpow, one_zpow]` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem Even.neg_pow : Even n → ∀ a : α, (-a) ^ n = a ^ n := by` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem even_neg_two : Even (-2 : α) := by simp only [even_neg, even_two]` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem Even.neg_zpow : Even n → ∀ a : α, (-a) ^ n = a ^ n := by` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem Even.odd_add : Even a → Odd b → Odd (b + a) :=` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem even_ofMul_iff {a : α} : Even (Additive.ofMul a) ↔ IsSquare a := Iff.rfl` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem Even.pow_of_ne_zero (ha : Even a) : ∀ {n : ℕ}, n ≠ 0 → Even (a ^ n)` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem Even.sub_odd (ha : Even a) (hb : Odd b) : Odd (a - b) := by` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem even_toAdd_iff {a : Multiplicative α} :` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem Even.trans_dvd (ha : Even a) (hab : a ∣ b) : Even b :=` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem Even.tsub [CanonicallyLinearOrderedAddCommMonoid α] [Sub α] [OrderedSub α]` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem even_two : Even (2 : α) :=` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem even_two_mul (a : α) : Even (2 * a) :=` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem IsSquare.div [DivisionCommMonoid α] {a b : α} (ha : IsSquare a) (hb : IsSquare b) :` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem isSquare_iff_exists_sq (m : α) : IsSquare m ↔ ∃ c, m = c ^ 2 := by simp [IsSquare, pow_two]` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem isSquare_inv : IsSquare a⁻¹ ↔ IsSquare a := by` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem IsSquare.map [MulOneClass α] [MulOneClass β] [FunLike F α β] [MonoidHomClass F α β]` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem IsSquare.mul [CommSemigroup α] {a b : α} : IsSquare a → IsSquare b → IsSquare (a * b) := by` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem isSquare_mul_self (m : α) : IsSquare (m * m) :=` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem isSquare_ofAdd_iff {a : α} : IsSquare (Multiplicative.ofAdd a) ↔ Even a := Iff.rfl` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem isSquare_one [MulOneClass α] : IsSquare (1 : α) :=` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem isSquare_op_iff {a : α} : IsSquare (op a) ↔ IsSquare a :=` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem IsSquare.pow (n : ℕ) : IsSquare a → IsSquare (a ^ n) := by` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem IsSquare_sq (a : α) : IsSquare (a ^ 2) :=` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem isSquare_toMul_iff {a : Additive α} : IsSquare (Additive.toMul a) ↔ Even a := Iff.rfl` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem isSquare_unop_iff {a : αᵐᵒᵖ} : IsSquare (unop a) ↔ IsSquare a := isSquare_op_iff.symm` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem isSquare_zero [MulZeroClass α] : IsSquare (0 : α) :=` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem IsSquare.zpow (n : ℤ) : IsSquare a → IsSquare (a ^ n) := by` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem odd_abs [LinearOrder α] : Odd (abs a) ↔ Odd a := by` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem Odd.add_even (ha : Odd a) (hb : Even b) : Odd (a + b) := by` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem Odd.add_odd : Odd a → Odd b → Even (a + b) := by` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem odd_iff_exists_bit1 : Odd a ↔ ∃ b, a = bit1 b :=` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem Odd.map [FunLike F α β] [RingHomClass F α β] (f : F) : Odd a → Odd (f a) := by` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem Odd.mul : Odd a → Odd b → Odd (a * b) := by` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem Odd.neg (ha : Odd a) : Odd (-a) := by` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem odd_neg : Odd (-a) ↔ Odd a :=` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem odd_neg_one : Odd (-1 : α) := by simp` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem Odd.neg_one_pow (h : Odd n) : (-1 : α) ^ n = -1 := by rw [h.neg_pow, one_pow]` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem Odd.neg_pow : Odd n → ∀ a : α, (-a) ^ n = -a ^ n := by` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem odd_one : Odd (1 : α) :=` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem Odd.pos [Nontrivial α] {a : α} (hn : Odd a) : 0 < a := by` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem Odd.pow (ha : Odd a) : ∀ {n : ℕ}, Odd (a ^ n)` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem Odd.sub_even (ha : Odd a) (hb : Even b) : Odd (a - b) := by` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem Odd.sub_odd (ha : Odd a) (hb : Odd b) : Even (a - b) := by` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem odd_two_mul_add_one (a : α) : Odd (2 * a + 1) :=` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem range_two_mul_add_one (α : Type*) [Semiring α] :` |
| a/Mathlib/Algebra/Parity.lean | `-` | `theorem range_two_mul (α) [Semiring α] : (Set.range fun x : α => 2 * x) = { a | Even a } := by` |
---
* 41  added declarations
* 62  removed declarations
* 2  paired declarations
---

You can run this locally using
```bash
./scripts/no_lost_declarations.sh <optional_commit>
```

**Comment by jcommelin** [comment]
My main takeaway is that the number of `+`s and `-`s is not the same.

**Comment by YaelDillies** [comment]
Yes, as I said in the PR description, I deleted some long deprecated lemmas that didn't really have a nice home anymore

**Comment by adomani** [comment]
In case it is helpful, I think that these are the declarations whose names are not paired:
```
+ Additive.instDecidablePredEven
- even_abs
- even_iff_exists_bit0
- Even.isSquare_pow
- Even.mul_left
- Even.mul_right
- Even.neg_pow
- even_two
- isSquare_inv
- isSquare_mul_self
- isSquare_ofAdd_iff
- IsSquare.pow
- IsSquare_sq
- isSquare_zero
- IsSquare.zpow
+ Multiplicative.instDecidablePredIsSquare
- odd_abs
- Odd.mul
- odd_neg
- Odd.neg_one_pow
- odd_one
- range_two_mul
- range_two_mul_add_one
```

**Comment by YaelDillies** [comment]
This list looks very wrong

**Comment by adomani** [comment]
I think that I know what is wrong: let me try to come up with something better!

**Comment by adomani** [comment]
@YaelDillies, I updated the list above: does it look better now?

**Comment by YaelDillies** [comment]
Since Johan complained, I've split the moves to #12829.

**Pull request closed without merging**